subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/DefaultConfigurationMetadata.java

```diff
@@ -291,6 +293,18 @@ Builder withCapabilities(CapabilitiesMetadata capabilities) {
             return this;
         }
 
+        /**
+         * When synthesizing variants, invoking this method prevents "fallback" to the default artifact of a maven dependency.
+         *
+         * <p>See {@link ConfigurationMetadata#requiresMavenArtifactDiscovery()}</p>
+         *
+         * @return this
+         */
+        Builder requiresMavenArtifactDiscovery() {
+            this.requiresMavenArtifactDiscovery = false;
+            return this;
+        }
+
```

Comment:
Let's remove this from the builder and instead set `requiresMavenArtifactDiscovery = false` when artifacts are set in `withArtifacts`.  I think that's the meaning of this flag (does the variant know what the artifacts are or do we need to look at the POM)